### DISCUSSION TITLE
Volume normalization

### DISCUSSION THREAD

**Discussion by JustCauseWhyNot** [wwmm/easyeffects#2477] [Q&A] [answered]
What would be the best way to normalize the volume in a game. Would it be using a compressor and then cranking the volume up, or is there better way? I'm having issues with a game where effect sounds like gun shots, and explosions are very loud while having foot steps very quiet. There are also no in game sliders to modify the sound beyond master volume.

**Comment by wwmm** [comment, marked as answer]
Usually for games I use the `compressor` and also the `autogain` plugin.

**Reply by JustCauseWhyNot** [reply]
I'll try out the autogain feature. What are the typical settings you use? And I'm assuming you use them both together at the same time right?

**Comment by wwmm** [comment]
>What are the typical settings you use?

The default autogain settings are usually enough for my needs. Depending on what kind of contents you are dealing with you may need to change the `maximum history` and the `reference` a little.

The compressor is the harder one to get right. If its threshold is too far above the current signal level it may not help much. One way to mitigate this problem is putting the autogain before it in the pipeline to make the average level to stay around the compressor threshold most of the time.

>And I'm assuming you use them both together at the same time right?

Yes.

Something else you may try is using the `maximizer` plugin instead of the compressor. In one of my presets I put the maximizer after the autogain. By playing with the maximizer `threshold` you will also be able to reduce the signal dynamic range.

**Reply by JustCauseWhyNot** [reply]
Thank you very much for the information. This is very helpful.

**Reply by JustCauseWhyNot** [reply]
Why would I want to use maximizer over compressor or vice versa? I really don't know which one would be better either in general or my needs.

**Reply by JustCauseWhyNot** [reply]
Also why would I not want maximum history to be set to something like 3m? Why is it the way it is, or would I want to use integrated reference?

**Comment by wwmm** [comment]
>Why would I want to use maximizer over compressor or vice versa?

As I am not an audio engineer or a musician I am not sure about the general answer. But from my point of view one key difference in favor of the compressor is flexibility. As you can see in its window there are lots of parameters as well as extra functionality like the `side chain` that the maximizer plugin does not offer. On the other hand the maximizer can also work like a limiter plugin. Something that the compressor can not do.

>Also why would I not want maximum history to be set to something like 3m?

The maximum history affects the `integrated loudness`. This is the loudness that is taking into account everything that was played within the maximum history window. The time window for the momentary loudness is of 400 milliseconds and for the short-term 3 seconds. These ones can not be changed.

The default reference mode takes the geometric average of these loudness values to estimate the effective loudness. So if you make the maximum history too long the integrated value may be taking into account data is not representative of the signal currently being played. And as a result the correction gain may not be adequate.

On the other hand if the integrated loudness history is too short the effective loudness will be dominated by the ones calculated in a shorter time window. And as a result the correction gain may be adjusted so fast that the corresponding variations in volume will be noticeable. It will depend on what is being played and on what you want to happen to the average sound volume.

**Reply by JustCauseWhyNot** [reply]
Thank you for the answers.

**Comment by JustCauseWhyNot** [comment]
I've got one last question. Should I expect to hear 2 different programs simultaneously with this effect stack eq->autogain->limiter->maximizer?

**Reply by wwmm** [reply]
EasyEffects has only one output effects pipeline. So if you enable effects for Firefox and mpv both programs outputs will be redirected to our virtual device. Inside this virtual device PipeWire will mix their audio content. The combination of both signals is what EasyEffects will apply effects to.

**Reply by JustCauseWhyNot** [reply]
That's helpful to know, but I think I worded my question incorrectly. I'm trying make 2 programs match each other in volume, and also have the quietest parts of a program = the loudest. Ik I've asked a lot of questions from you, but do know I've read the manual as much as possible in order to answer my questions. Thank you for the program, and your time.

**Reply by wwmm** [reply]
>I'm trying make 2 programs match each other in volume, and also have the quietest parts of a program = the loudest

I see. I am not sure if it is possible to achieve this using only EasyEffects. Exactly because of the explanation I gave before. We redirect all audio streams to the same device and mix everything because we have only one output effects pipeline. So the autogain, the compressor and the other plugins will act on the combined signal and not individually on each stream data.

**Reply by JustCauseWhyNot** [reply]
That's very good to know. So it's not crazy to also adjust the volume in pavu after the effects are applied. Thank you very much!

**Reply by wwmm** [reply]
It will probably be better than not using EasyEffects. But not the same as processing each stream individually and forcing the average signal strength match.

**Reply by JustCauseWhyNot** [reply]
Is there any way to?
>processing each stream individually and forcing the average signal strength match.

**Reply by wwmm** [reply]
>Is there any way to?

Unfortunately no. One pipeline for each app makes things a lot more complicated to handle. And our window would have to go through lots of changes to enable this to be configured.

**Reply by JustCauseWhyNot** [reply]
Alright. Thank you.
